projects/ngrx.io/content/guide/store/architecture.md

```diff
@@ -1 +1,418 @@
+
+  constructor(public payload: { username: string; password: string }) {}
+}
+</code-example>
+
+We've discussed previously how the above code, using Typescript, provides a lot of benefits over creating Actions using plain JavaScript and object literals or even using classes. We can take this one step further though, by using enums. Right now our `type` is a string. Let's first create an enum like so:
+
+<code-example header="An ActionTypes enum">
+export enum ActionTypes {
+  Login = '[Login Page] Login',
+}
+</code-example>
+As you can see above we've moved our `type` value into the `ActionTypes` enum and now we can perform the next step which is to replace our `type` value with our enum value:
+
+<code-example header="login-page.actions.ts">
+import { Action } from '@ngrx/store';
+
+export enum ActionTypes {
+  Login = '[Login Page] Login',
+}
+
+export class Login implements Action {
+  readonly type = ActionTypes.Login;
+
+  constructor(public payload: { username: string; password: string }) {}
+}
+</code-example>
+
+Now this scales. Imagine we need a new action `Logout`. That class can use the same enum. The `Logout` class would look like so:
+
+<code-example header="login-page.actions.ts">
+import { Action } from '@ngrx/store';
+
+export enum ActionTypes {
+  Login = '[Login Page] Login',
+  Logout = '[Login Page] Logout'
+}
+
+export class Logout implements Action {
+  readonly type = ActionTypes.Logout
+}
+
+export class Login implements Action {
+  readonly type = ActionTypes.Login;
+
+  constructor(public payload: { username: string; password: string }) {}
+}
+</code-example>
+
+Above we've added the `Logout` type to our `ActionTypes` enum and we've also created our `Logout` action class. As you can see everything that goes together can be using the same enum.
+
+### Union types
+Ok, so we've come to a point where we have defined two different actions `Login` and `Logout`. Let's see what an accompanying reducer could look like:
+
+<code-example header="Login, Logout reducer">
+function reducer(state, action) {
+  switch(action.type) {
+    // define code here
+  }
+}
+</code-example>
+
+As you can see above we have no typesafety. We could be feeding this `reducer()` function any kind of parameters. So lets fix that shall we ?
+The first thing we are going to do is add something called a union type. Union types works in the following way:
+
+<code-example header="Example union">
+type Cars = 'Ferrari' | 'Wolkswagen' | 'Volvo';
+
+function car(car: Cars) {
+  console.log(car);
+}
+
+car('Ferrari'); // OK
+car('Porsche'); // Not OK
+car('Volvo') // OK
+</code-example>
+
+As you can see from above code we are defining the type `Cars` and what we are doing is adding a number of values that it can assume. Each value it can assume are separated by a pipe, `|`. The example invocation of the method `car()` tells us that `Ferrari` and `Volvo` are ok to add but not `Porsche` as it's not part of the definition of our union type.
```

Comment:
```suggestion
As you can see from the above code we are defining the type `Cars` and what we are doing is adding a number of values that it can assume. Each value it can assume are separated by a pipe, `|`. The example invocation of the method `car()` tells us that `Ferrari` and `Volvo` are ok to add but not `Porsche` as it's not part of the definition of our union type.
```